### PR TITLE
Initial push of Harman Luxury Driver

### PR DESCRIPTION
**Pull request by alon-tchelet**
Initial push of harman-luxury SmartThings edge driver.
Current version supports only future released device L42ms

### CODE REVIEW

File: drivers/SmartThings/harman-luxury/profiles/stream-sdk.yaml
Comment (FreeMasen) [inline drivers/SmartThings/harman-luxury/profiles/stream-sdk.yaml:32]: nit: commented code

File: drivers/SmartThings/harman-luxury/profiles/stream-sdk.yaml
Comment (FreeMasen) [inline drivers/SmartThings/harman-luxury/profiles/stream-sdk.yaml:36]: nit: missing trailing new line

File: drivers/SmartThings/harman-luxury/src/api/nsdk.lua
Comment (FreeMasen) [inline drivers/SmartThings/harman-luxury/src/api/nsdk.lua:41]:
```suggestion
        local t = {}
        for i, role in rolesList do
            if roles == "all" then
                return "@all"
            else
                table.insert(t, role)
            end
        end
        return table.concat(t, ",")
```

File: drivers/SmartThings/harman-luxury/src/api/nsdk.lua
Comment (FreeMasen) [inline drivers/SmartThings/harman-luxury/src/api/nsdk.lua:66]:
```suggestion
        return json.decode(table.concat(sink, ""))[1]
```

File: drivers/SmartThings/harman-luxury/src/api/nsdk.lua
Comment (FreeMasen) [inline drivers/SmartThings/harman-luxury/src/api/nsdk.lua:99]: again, `sink` should be concatenated

**Review by FreeMasen — COMMENTED** [review]
Quick look, there is some inconsistency that should probably be addressed but mostly nits.

- All uses of the `sink` in `http.request` should be updated to use `table.concat(sink, "")` to avoid missing data if it comes across as multiple chunks.
- Indentation should be consistent, some files are 8 spaces, some are 4 (I think more are 4 than 8)

File: drivers/SmartThings/harman-luxury/src/api/apis.lua
Comment (greens) [inline drivers/SmartThings/harman-luxury/src/api/apis.lua:52]: typo: `Manufacturer`

File: drivers/SmartThings/harman-luxury/src/disco_helper.lua
Comment (greens) [inline drivers/SmartThings/harman-luxury/src/disco_helper.lua:88]: these should be `elseif`s I think

File: drivers/SmartThings/harman-luxury/src/handlers.lua
Comment (greens) [inline drivers/SmartThings/harman-luxury/src/handlers.lua:191]: this should either just call `handle_stop` and `handle_playPause` or they should call it, to avoid unnecessary code duplication

File: drivers/SmartThings/harman-luxury/src/handlers.lua
Comment (greens) [inline drivers/SmartThings/harman-luxury/src/handlers.lua:83]: again, since you've already defined mute and unmute methods, this method seems reduplicative

**Review by greens — COMMENTED** [review]
Please fix all of the linter errors.

Additionally, a lot of the handlers have significant amount of duplicated code. A better pattern would be worthwhile, I think. Either combining methods like on/off and mute/unmute or using a factory would be better I think.

There are also a significant number of magic strings (like `"device_ipv4"`) that I think would be good to define somewhere, maybe in a constants file.

File: drivers/SmartThings/harman-luxury/search-parameters.yml
Comment (dljsjr) [inline drivers/SmartThings/harman-luxury/search-parameters.yml:2]: Is this the correct service name? It's not the same one we saw in test tools, so just wanted to make sure this was accurate.

File: drivers/SmartThings/harman-luxury/src/disco_helper.lua
Comment (dljsjr) [inline drivers/SmartThings/harman-luxury/src/disco_helper.lua:126]: type: `servie` should be `service`

File: drivers/SmartThings/harman-luxury/src/disco_helper.lua
Comment (dljsjr) [inline drivers/SmartThings/harman-luxury/src/disco_helper.lua:140]: Just making a note: This is the correct structure for the mDNS API; released versions of the API will always use the `.found` field.

The rest of the code should treat the `{ answers, additional }` form as the fallback, not the `found` form.

File: drivers/SmartThings/harman-luxury/src/devices.lua
Comment (dljsjr) [inline drivers/SmartThings/harman-luxury/src/devices.lua:8]: typ: `suppourted` -> `supported`

File: drivers/SmartThings/harman-luxury/src/devices.lua
Comment (dljsjr) [inline drivers/SmartThings/harman-luxury/src/devices.lua:27]: Since we want the values here to not only be non-nil, but also explicitly `strings`, it would be more defensive and less bug-prone to use `if type(var) ~= "string"` instead of `if not var`.

So for example, `if type(label) ~= "string"` instead of `if not label`. This check will catch both `nil` and cases where the value passed in is not a string.

File: drivers/SmartThings/harman-luxury/src/devices.lua
Comment (dljsjr) [inline drivers/SmartThings/harman-luxury/src/devices.lua:122]: Similar to the comment that @greens made elsewhere, there's a lot of duplicated code here. One option would be to have Lua tables that take `params.setupid` as the key, and return the MFR and Model information for the various speakers, as those are the only things that differ in the various functions defined in this file.

File: drivers/SmartThings/harman-luxury/src/disco.lua
Comment (dljsjr) [inline drivers/SmartThings/harman-luxury/src/disco.lua:11]: This one is a nit:

This file (and a few others) use a lot of string concatenation (the `..` operator).

`..` is fragile; it will crash on `nil` values if `tostring` isn't called first. Most of the usage here is well protected, so that's good.

However, the overall readability when there are multiple interleaved/concatenated variables in the middle of a string suffers, and the use of `tostring` to elide `nil` isn't very obvious.

In cases where one needs to insert multiple variables in to a string, it can often lead to much more readable and maintainble logging if `string.format()` is used instead of multiple concatenation operations.

File: drivers/SmartThings/harman-luxury/src/disco.lua
Comment (dljsjr) [inline drivers/SmartThings/harman-luxury/src/disco.lua:42]: Already mentioned elsewhere but it would be nice if this string (`"device_ipv4`) was captured in a constant somewhere.

File: drivers/SmartThings/harman-luxury/src/disco_helper.lua
Comment (dljsjr) [inline drivers/SmartThings/harman-luxury/src/disco_helper.lua:149]: This is correct.

I think it's good that there is code here that catches both the `{answers, additional}` form and the `found` form, but style-wise, the code should prefer the `found` form.

File: drivers/SmartThings/harman-luxury/src/handlers.lua
Comment (dljsjr) [inline drivers/SmartThings/harman-luxury/src/handlers.lua:5]: Seconding @greens comment that there's a lot of duplicated code in here.

File: drivers/SmartThings/harman-luxury/src/api/gets.lua
Comment (dljsjr) [inline drivers/SmartThings/harman-luxury/src/api/gets.lua:19]: I'd recommend `if type(ret) == "table"` instead.

File: drivers/SmartThings/harman-luxury/src/api/gets.lua
Comment (dljsjr) [inline drivers/SmartThings/harman-luxury/src/api/gets.lua:29]: Unused function.

File: drivers/SmartThings/harman-luxury/src/api/nsdk.lua
Comment (dljsjr) [inline drivers/SmartThings/harman-luxury/src/api/nsdk.lua:5]: `json.decode` can throw an `error` on malformed input, which can crash the whole driver.

This API needs to be either documented to note this, or cleaned up to avoid this. In the meantime, I would strongly suggest using `pcall` everywhere that you call `json.decode`

**Review by dljsjr — CHANGES_REQUESTED** [review]
I made a lot of comments, but they can be boiled down to a handful of things that need to be addressed before we can merge.

1. Luacheck lints are failing. Luacheck can be installed from Luarocks and can be run with the config file vendored in this repo's .github directory for validating locally without needing to wait for the PR actions to complete.
2. Lots of code duplication that should be cleaned up,
3. The discovery code that consumes the `mDNS` API is technically valid but has a preference for a deprecated API, and it treats the stable API as the "fallback". For style, readability, and maintainability reasons that should be switched.
4. Weak error handling, that can be broken down even further

### Error Handling Improvement Requests

#### JSON Error Handling

`json.decode` needs to be wrapped in `pcall` to avoid accidentally crashing the entire driver if there's malformed input. You can find examples of this in the up to date Hue and Sonos drivers.

#### Function Input and Return Value Validation

There are a lot of places where the return values from network I/O operations are checked for "truthiness" to see if the values are `nil`. Network operations are fragile, and various forms of socket interruptions can lead to values that might not be what one would expect.

When checking the type of a value that has come directly from a socket I/O operation, or when checking that a value is safe to forward to another function, it's safest to use `type(val) == "string"` or whatever the expected type string is. Because `nil` is its own type, this satisfies both nil checking and rudimentary argument type checking needs.

#### API Error Handling

The `apis` directory and all of its modules have very incosistent error handling patterns. In many places they return internal API calls directly, such as `return json.decode(…)`, but in their local error checking branches they will often swallow errors entirely or log them in-flight then return `false`.

This is both diverging from the return semantics of the internal APIs that it passes through, as well as not being idiomatic Lua. It also puts undue burden on the caller to know that the return semantics of those modules can be very polymorphic, and make the code hard to reason about.

A more reliable error handling pattern would be to return `nil, "error message string"` when a function fails instead of logging and returning only `false`. This does a few things:

1. Less code specialization in the function bodies so that there can be more opportunities for code reuse
2. Makes it easier to write functions using the early return pattern, which we prefer
3. Makes the code more predictable, as this is a common pattern in most of the Lua libraries that we use so it makes the code behave more consistently withour dependencies and platform-provided libraries
5. Using `nil` instead of `false` removes any ambiguity from a function that could potentially have a `bool` return value that can be false in the non-error case; even if that type of function isn't in this codebase, it eliminates the question of whether or not it could be from a reader/maintainer.

File: drivers/SmartThings/harman-luxury/search-parameters.yml
Comment (alon-tchelet) [reply drivers/SmartThings/harman-luxury/search-parameters.yml:2]: Yes. We changed it one of our latest firmware releases. When you test on the L42ms, please ensure you are testing on latest StreamSDK software. Latest release currently is 1793, but we will make a new release once I covered the changes.

Reason for the change was to avoid discovery of unsupported devices

**Comment by alon-tchelet** [comment]
@dljsjr Hi, Thanks for all the comments. Please have a look at latest master after I addressed all the previous comments

File: drivers/SmartThings/harman-luxury/src/api/nsdk.lua
Comment (greens) [inline drivers/SmartThings/harman-luxury/src/api/nsdk.lua:1]: you are repeating code a _lot_ in this file

**Review by greens — COMMENTED** [review]
never mind, I see it

**Review by greens — CHANGES_REQUESTED** [review]
We use 2-width spaces in most of our drivers. You've got a mix of tabs and some 4-width spaces here. Could you update that for consistency?

File: drivers/SmartThings/harman-luxury/src/handlers.lua
Comment (alon-tchelet) [reply drivers/SmartThings/harman-luxury/src/handlers.lua:64]: Absolutely, correct. Was just looking at this after some tests results

**Comment by alon-tchelet** [comment]
Hi again, I applied the changes you requested (fixed indents and further minimized repeated code inside nsdk.lua). 
Once again ready for review

File: drivers/SmartThings/harman-luxury/config.yaml
Comment (dljsjr) [inline drivers/SmartThings/harman-luxury/config.yaml:1]: This key should be a "human friendly" name, and as such doesn't need the hyphen.

File: drivers/SmartThings/harman-luxury/src/api/nsdk.lua
Comment (dljsjr) [inline drivers/SmartThings/harman-luxury/src/api/nsdk.lua:82]: Returning the pcall's `ret` value is basically the same thing as re-throwing an exception. I would prefer if the pcall-specific error handling was internal to the function, and I would also prefer to stick with the Lua convention of returning the value or `nil` in the first return spot, and `nil` or an error message in the 2nd return spot.

File: drivers/SmartThings/harman-luxury/src/init.lua
Comment (dljsjr) [inline drivers/SmartThings/harman-luxury/src/init.lua:317]: See https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/pull/885#discussion_r1317431712

File: drivers/SmartThings/harman-luxury/src/init.lua
Comment (dljsjr) [inline drivers/SmartThings/harman-luxury/src/init.lua:238]: There is non-trivial overhead to creating many tasks; in the case where a user has many speakers in their home, there could be a lot of overhead with having a task per device. I would recommend creating a single background task that monitors for IP address changes and then fans out to the appropriate device.

File: drivers/SmartThings/harman-luxury/src/api/nsdk.lua
Comment (dljsjr) [inline drivers/SmartThings/harman-luxury/src/api/nsdk.lua:67]: Please see @FreeMasen's comment here: https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/pull/885#pullrequestreview-1565972963

File: drivers/SmartThings/harman-luxury/src/api/nsdk.lua
Comment (dljsjr) [inline drivers/SmartThings/harman-luxury/src/api/nsdk.lua:87]: Same here, and everywhere else `pcall` is used; I would prefer to see `pcall` "catches" handled inline instead of being bubbled up.

Just to reiterate: I think it's okay that these functions are fallible. But I would like to see internal error handling with a relevant error based on the `pcall` failure reported in the returns, instead of just bubbling up the boolean transparently.

File: drivers/SmartThings/harman-luxury/src/api/nsdk.lua
Comment (dljsjr) [reply drivers/SmartThings/harman-luxury/src/api/nsdk.lua:1]: Second this, lots of opportunities for "DRY" refactors in this file I'd like to see seized.

**Comment by alon-tchelet** [comment]
Covered all comments again. Hope it's better now

File: drivers/SmartThings/harman-luxury/src/api/apis.lua
Comment (dljsjr) [reply drivers/SmartThings/harman-luxury/src/api/apis.lua:56]: Typo remains; `Manufature` should be `Manufacturer`

**Review by dljsjr — APPROVED** [review]
There are a few remaining style nits but this looks pretty good to me at this point, have there been tests with real devices and real hubs?

**Review by dljsjr — APPROVED** [review]
@alon-tchelet Looks ready to go, but this needs to be rebased on the usptream `main` before we can merge.

**Review by greens — APPROVED** [review]

**Pull request merged**
